### PR TITLE
cls/rgw: index cancelation still cleans up remove_objs

### PR DESCRIPTION
**Pull request by cbodley**
when multipart uploads complete their final bucket index transaction, they pass the list of part objects in 'remove_objs' for bulk removal - the part objects, along with their bucket stats, get replaced by the head object

but if CompleteMultipart races with another upload, the head object write will fail with ECANCELED and the bucket index transaction gets canceled with CLS_RGW_OP_CANCEL. these canceled uploads still need to clean up their 'remove_objs', but cancelation was returning too early. as a result, these bucket index entries get orphaned and leave the bucket stats inconsistent

this commit reworks rgw_bucket_complete_op() so that CLS_RGW_OP_CANCEL is handled the same way as OP_ADD and OP_DEL, so always runs the loop to clean up 'remove_objs'

Fixes: https://tracker.ceph.com/issues/53199

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cbodley** [comment]
with this fix, the bucket stats correctly show a single 32M object using the reproducer from https://tracker.ceph.com/issues/53199, even with 36 concurrent multipart uploads (30 of which were canceled):
```
    "usage": {
        "rgw.main": {
            "size": 33554432,
            "size_actual": 33554432,
            "size_utilized": 33554432,
            "size_kb": 32768,
            "size_kb_actual": 32768,
            "size_kb_utilized": 32768,
            "num_objects": 1
        },
        "rgw.multimeta": {
            "size": 0,
            "size_actual": 0,
            "size_utilized": 0,
            "size_kb": 0,
            "size_kb_actual": 0,
            "size_kb_utilized": 0,
            "num_objects": 0
        }
```

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -990,6 +990,42 @@ static int read_key_entry(cls_method_context_t hctx, cls_rgw_obj_key& key,
...
 
+static int complete_cleanup_obj(cls_method_context_t hctx,
+                                rgw_bucket_dir_header& header,
+                                const cls_rgw_obj_key& key, bool log_op)
+{
+  rgw_bucket_dir_entry entry;
+  string idx;
+  int ret = read_key_entry(hctx, key, &idx, &entry);
+  if (ret < 0) {
+    CLS_LOG(1, "%s: read_index_entry name=%s instance=%s failed with %d",
```
Comment (ivancich) [inline src/cls/rgw/cls_rgw.cc:1001]: literal string should contain "read_key_entry", I think

File: src/cls/rgw/cls_rgw.cc [hunk 2]
```diff
@@ -990,6 +990,42 @@ static int read_key_entry(cls_method_context_t hctx, cls_rgw_obj_key& key,
   return 0;
 }
 
+static int complete_cleanup_obj(cls_method_context_t hctx,
```
Comment (ivancich) [inline src/cls/rgw/cls_rgw.cc:993]: I would love to see a high-level comment describing the purpose/role of this function

File: src/cls/rgw/cls_rgw.cc [hunk 3]
```diff
@@ -1043,113 +1079,89 @@ int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist
...
-    }
-
-    ret = cls_cxx_map_remove_key(hctx, k);
-    if (ret < 0) {
-      CLS_LOG(1, "rgw_bucket_complete_op(): cls_cxx_map_remove_key, failed to remove entry, name=%s instance=%s read_index_entry ret=%d", remove_key.name.c_str(), remove_key.instance.c_str(), rc);
-      continue;
+  for (const auto& remove_key : op.remove_objs) {
+    rc = complete_cleanup_obj(hctx, header, remove_key, log_op);
+    if (rc < 0) {
+      continue; // part cleanup errors are not fatal
```
Comment (ivancich) [inline src/cls/rgw/cls_rgw.cc:1164]: even though this logic doesn't change behavior, I like that it's here to document the decision made

**Review by ivancich — COMMENTED** [review]
nice!

**Review by ivancich — APPROVED** [review]
oops, my previous review was not the approve that I intended

**Comment by cbodley** [comment]
thanks for the review @ivancich! i'm still planning to model multipart uploads in https://github.com/ceph/ceph/pull/43843, so that should give us good test coverage for this

**Comment by cbodley** [comment]
rebased over https://github.com/ceph/ceph/pull/43103

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test api

**Pull request merged**